runners/apex/src/main/java/org/apache/beam/runners/apex/translation/operators/ApexTimerInternals.java

```diff
@@ -147,55 +161,106 @@ public Instant currentOutputWatermarkTime() {
-    final Slice keyBytes;
-    try {
-      keyBytes = new Slice(CoderUtils.encodeToByteArray(keyCoder, key));
-    } catch (CoderException e) {
-      throw new RuntimeException(e);
-    }
-    Set<Slice> timersForKey = activeTimers.get(keyBytes);
-    if (timersForKey == null) {
-      timersForKey = new HashSet<>();
+    // fire ready timers
+    if (!toFire.isEmpty()) {
+      for (Slice keyBytes : toFire.keySet()) {
+        try {
+          K key = CoderUtils.decodeFromByteArray(keyCoder, keyBytes.buffer);
+          timerProcessor.fireTimer(key, toFire.get(keyBytes));
+        } catch (CoderException e) {
+          throw new RuntimeException(e);
+        }
+      }
     }
+  }
 
+  private Slice getKeyBytes(K key) {
     try {
-      Slice timerBytes = new Slice(CoderUtils.encodeToByteArray(timerDataCoder, timer));
-      timersForKey.add(timerBytes);
+      return new Slice(CoderUtils.encodeToByteArray(keyCoder, key));
     } catch (CoderException e) {
       throw new RuntimeException(e);
     }
-
-    activeTimers.put(keyBytes, timersForKey);
   }
 
-  private void unregisterActiveTimer(K key, TimerData timer) {
-    final Slice keyBytes;
-    try {
-      keyBytes = new Slice(CoderUtils.encodeToByteArray(keyCoder, key));
-    } catch (CoderException e) {
-      throw new RuntimeException(e);
+  protected static class TimerSet implements Serializable {
+    private final Map<Slice, Set<Slice>> activeTimers = new HashMap<>();
+    private final TimerDataCoder timerDataCoder;
+
+    protected TimerSet(TimerDataCoder timerDataCoder) {
+      this.timerDataCoder = timerDataCoder;
     }
 
-    Set<Slice> timersForKey = activeTimers.get(keyBytes);
-    if (timersForKey != null) {
+    public void addTimer(Slice keyBytes, TimerData timer) {
+      Set<Slice> timersForKey = activeTimers.get(keyBytes);
+      if (timersForKey == null) {
+        timersForKey = new HashSet<>();
+      }
+
       try {
         Slice timerBytes = new Slice(CoderUtils.encodeToByteArray(timerDataCoder, timer));
         timersForKey.add(timerBytes);
-        timersForKey.remove(timerBytes);
       } catch (CoderException e) {
         throw new RuntimeException(e);
       }
 
+      activeTimers.put(keyBytes, timersForKey);
+    }
+
+    public void deleteTimer(Slice keyBytes, StateNamespace namespace, String timerId) {
+      Set<Slice> timersForKey = activeTimers.get(keyBytes);
+      if (timersForKey == null) {
+        return;
+      }
+
+      Iterator<Slice> timerIt = timersForKey.iterator();
+      while (timerIt.hasNext()) {
+        try {
+          TimerData timerData = CoderUtils.decodeFromByteArray(timerDataCoder,
+              timerIt.next().buffer);
+          ComparisonChain chain =
+              ComparisonChain.start().compare(timerData.getTimerId(), timerId);
+          if (chain.result() == 0 && !timerData.getNamespace().equals(namespace)) {
+            // Obtaining the stringKey may be expensive; only do so if required
```

Comment:
Well, hmm. I guess keep it that way since there is probably some context for it. Perhaps it was a workaround for [BEAM-1022](https://issues.apache.org/jira/browse/BEAM-1022) that was recently fixed.